include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,447 @@
+#define OP_PLASTIC			2u
+#define OP_COATING			3u
+#define OP_MAX_BRDF			OP_COATING
+#define OP_DIFFTRANS		4u
+#define OP_DIELECTRIC		5u
+#define OP_MAX_BSDF			OP_DIELECTRIC
+#define OP_BLEND			6u
+#define OP_BUMPMAP			7u
+#define OP_SET_GEOM_NORMAL	8u
+#define OP_INVALID			9u
+#define OutColor			10u
+
+#define NDF_BECKMANN	0u
+#define NDF_GGX			1u
+#define NDF_PHONG		2u
+#define NDF_AS			3u
+
+#include <irr/builtin/glsl/bxdf/common.glsl>
+#include <irr/builtin/glsl/bxdf/brdf/specular/fresnel/fresnel.glsl>
+#include <irr/builtin/glsl/bxdf/brdf/diffuse/fresnel_correction.glsl>
+#include <irr/builtin/glsl/bxdf/brdf/diffuse/lambert.glsl>
+#include <irr/builtin/glsl/bxdf/brdf/diffuse/oren_nayar.glsl>
+#include <irr/builtin/glsl/bxdf/brdf/specular/ndf/ggx.glsl>
+#include <irr/builtin/glsl/bxdf/brdf/specular/ashikhmin_shirley.glsl>
+#include <irr/builtin/glsl/bxdf/brdf/specular/beckmann_smith.glsl>
+#include <irr/builtin/glsl/bxdf/brdf/specular/ggx.glsl>
+#include <irr/builtin/glsl/bxdf/brdf/specular/blinn_phong.glsl>
+#include <irr/builtin/glsl/bump_mapping/utils.glsl>
+
+irr_glsl_BSDFAnisotropicParams currBSDFParams;
+irr_glsl_AnisotropicViewSurfaceInteraction currInteraction;
+reg_t registers[REG_COUNT];
+reg_t prefetch_registers[TEX_PREFETCH_REG_COUNT];
+reg_t normal_registers[NORMAL_REG_COUNT];
+
+vec3 textureOrRGBconst(in uvec3 data, in bool texPresenceFlag)
+{
+	vec3 retval;
+	if (texPresenceFlag)
+		retval = prefetch_registers[data.z];
+	else
+		retval = uintBitsToFloat(data);
+	return retval;
+}
+float textureOrRconst(in uvec3 data, in bool texPresenceFlag)
+{
+	float retval;
+	if (texPresenceFlag)
+		retval = prefetch_registers[data.z].x;
+	else
+		retval = uintBitsToFloat(data.x);
+	return retval;
+}
+
+void setCurrBSDFParams(in vec3 n, in vec3 L)
+{
+	vec3 campos = irr_glsl_SBasicViewParameters_GetEyePos(CamData.params.NormalMatAndEyePos);
+	irr_glsl_IsotropicViewSurfaceInteraction interaction = irr_glsl_calcFragmentShaderSurfaceInteraction(campos, WorldPos, n);
+	currInteraction = irr_glsl_calcAnisotropicInteraction(interaction);
+	irr_glsl_BSDFIsotropicParams isoparams = irr_glsl_calcBSDFIsotropicParams(interaction, L);
+	currBSDFParams = irr_glsl_calcBSDFAnisotropicParams(isoparams, currInteraction);
+}
+
+void instr_execute_DIFFUSE(in instr_t instr, in uvec3 regs, in bsdf_data_t data)
+{
+	if (currBSDFParams.isotropic.NdotL>FLT_MIN)
+	{
+		vec3 refl = textureOrRGBconst(data.data[1].zw, instr_getReflectanceTexPresence(instr));
+		vec3 diffuse = irr_glsl_lambertian_cos_eval(currBSDFParams.isotropic,currInteraction.isotropic) * refl;
+		registers[REG_DST(regs)] = diffuse;
+	}
+	else registers[REG_DST(regs)] = reg_t(0.0);
+}
+void instr_execute_DIFFTRANS(in instr_t instr, in uvec3 regs, in mat2 dUV, in vec3 trans, in bsdf_data_t data)
+{
+	registers[REG_DST(regs)] = reg_t(1.0, 0.0, 0.0);
+}
+void instr_execute_DIELECTRIC(in instr_t instr, in uvec3 regs, in bsdf_data_t data)
+{
+	if (currBSDFParams.isotropic.NdotL>FLT_MIN)
```

Comment:
`if (currBSDFParams.isotropic.NdotL>FLT_MIN)` shall be hoisted outside of the `instr_execute_`

Reply:
for every function

Reply:
also put the `registers[REG_DST(regs)` outside as well (by making the `instr_execute` return a value

Reply:
done?

Reply:
TODO